File: packages/pixlive/ios/Plugin/Pixlive.swift
import Foundation
import UIKit
import Capacitor
import VDARSDK

// swiftlint:disable:next type_body_length
@objc public class Pixlive: NSObject, VDARSDKControllerDelegate, RemoteControllerDelegate {

    private weak var plugin: PixlivePlugin?
    private var arViewController: VDARLiveAnnotationViewController?
    private var touchForwarderView: TouchForwarderView?
    private var currentContext: VDARContext?

    public init(_ plugin: PixlivePlugin) {
        self.plugin = plugin
        super.init()
    }

    @objc public func initialize() {
        guard let plugin = plugin else { return }
        let licenseKey = plugin.getConfig().getString("licenseKey") ?? ""
        let apiUrl = plugin.getConfig().getString("apiUrl")
        let sdkUrl = plugin.getConfig().getString("sdkUrl")

        guard let documentsPath = NSSearchPathForDirectoriesInDomains(.documentDirectory, .userDomainMask, true).first else { return }
        let storagePath = (documentsPath as NSString).appendingPathComponent("pixliveSDK")
        let fileManager = FileManager.default
        if !fileManager.fileExists(atPath: storagePath) {
            try? fileManager.createDirectory(atPath: storagePath, withIntermediateDirectories: true)
        }

        VDARSDKController.startSDK(storagePath, withLicenseKey: licenseKey)
        guard let controller = VDARSDKController.sharedInstance() else { return }
        if let apiUrl = apiUrl {
            VDARRemoteController.sharedInstance()?.customAPIServer = apiUrl
        }
        if let sdkUrl = sdkUrl {
            VDARRemoteController.sharedInstance()?.customSdkServer = sdkUrl
        }
        controller.enableCodesRecognition = true
        let cameraSender = VDARCameraImageSource()
        controller.imageSender = cameraSender
        controller.detectionDelegates.add(self)
        VDARRemoteController.sharedInstance()?.delegate = self
    }

    @objc public func synchronize(_ options: SynchronizeOptions, completion: @escaping (_ error: Error?) -> Void) {
        let priors = PixliveHelper.buildTagPriors(options.tags)
        VDARRemoteController.sharedInstance()?.syncRemoteModelsAsynchronously(withPriors: priors, withCompletionBlock: { _, err in
            if let err = err {
                completion(err)
            } else {
                completion(nil)
            }
        })
    }

    @objc public func synchronizeWithToursAndContexts(_ options: SynchronizeWithToursAndContextsOptions, completion: @escaping (_ error: Error?) -> Void) {
        let priors = PixliveHelper.buildFullPriors(tags: options.tags, tourIds: options.tourIds, contextIds: options.contextIds)
        VDARRemoteController.sharedInstance()?.syncRemoteModelsAsynchronously(withPriors: priors, withCompletionBlock: { _, err in
            if let err = err {
                completion(err)
            } else {
                completion(nil)
            }
        })
    }

    @objc public func updateTagMapping(_ options: UpdateTagMappingOptions, completion: @escaping (_ error: Error?) -> Void) {
        VDARRemoteController.sharedInstance()?.syncTagContexts(options.tags, withCompletionBlock: { _, err in
            if let err = err {
                completion(err)
            } else {
                completion(nil)
            }
        })
    }

    @objc public func enableContextsWithTags(_ options: EnableContextsWithTagsOptions, completion: @escaping (_ error: Error?) -> Void) {
        VDARSDKController.sharedInstance()?.disableContexts()
        VDARSDKController.sharedInstance()?.enableContexts(withTags: options.tags)
        completion(nil)
    }

    @objc public func getContexts(completion: @escaping (_ result: GetContextsResult?, _ error: Error?) -> Void) {
        guard let controller = VDARSDKController.sharedInstance() else {
            completion(nil, CustomError.sdkNotInitialized)
            return
        }
        var contextsArray = JSArray()
        if let contextIds = controller.contextIDs as? [String] {
            for contextId in contextIds {
                if let context = controller.getContext(contextId) {
                    contextsArray.append(PixliveHelper.contextToJSObject(context))
                }
            }
        }
        completion(GetContextsResult(contexts: contextsArray), nil)
    }

    @objc public func getContext(_ options: GetContextOptions, completion: @escaping (_ result: GetContextResult?, _ error: Error?) -> Void) {
        guard let controller = VDARSDKController.sharedInstance() else {
            completion(nil, CustomError.sdkNotInitialized)
            return
        }
        guard let context = controller.getContext(options.contextId) else {
            completion(nil, CustomError.contextNotFound)
            return
        }
        completion(GetContextResult(context: PixliveHelper.contextToJSObject(context)), nil)
    }

    @objc public func activateContext(_ options: ActivateContextOptions, completion: @escaping (_ error: Error?) -> Void) {
        guard let controller = VDARSDKController.sharedInstance() else {
            completion(CustomError.sdkNotInitialized)
            return
        }
        if let context = controller.getContext(options.contextId) {
            context.activate()
        }
        completion(nil)
    }

    @objc public func stopContext(completion: @escaping (_ error: Error?) -> Void) {
        if let context = currentContext {
            context.stop()
        }
        completion(nil)
    }

    @objc public func getNearbyGPSPoints(_ options: GetNearbyGPSPointsOptions, completion: @escaping (_ result: GetNearbyGPSPointsResult?, _ error: Error?) -> Void) {
        guard let controller = VDARSDKController.sharedInstance() else {
            completion(nil, CustomError.sdkNotInitialized)
            return
        }
        let points = controller.getNearbyGPSPointsfromLat(options.latitude, lon: options.longitude)
        var pointsArray = JSArray()
        if let points = points {
            for point in points {
                pointsArray.append(PixliveHelper.gpsPointToJSObject(point))
            }
        }
        completion(GetNearbyGPSPointsResult(points: pointsArray), nil)
    }

    @objc public func getGPSPointsInBoundingBox(_ options: GetGPSPointsInBoundingBoxOptions, completion: @escaping (_ result: GetGPSPointsInBoundingBoxResult?, _ error: Error?) -> Void) {
        guard let controller = VDARSDKController.sharedInstance() else {
            completion(nil, CustomError.sdkNotInitialized)
            return
        }
        let points = controller.getGPSPoints(
            inBoundingBoxOfMinLat: options.minLatitude, minLon: options.minLongitude,
            maxLat: options.maxLatitude, maxLon: options.maxLongitude
        )
        var pointsArray = JSArray()
        if let points = points {
            for point in points {
                pointsArray.append(PixliveHelper.gpsPointToJSObject(point))
            }
        }
        completion(GetGPSPointsInBoundingBoxResult(points: pointsArray), nil)
    }

    @objc public func getNearbyBeacons(completion: @escaping (_ result: GetNearbyBeaconsResult?, _ error: Error?) -> Void) {
        guard let controller = VDARSDKController.sharedInstance() else {
            completion(nil, CustomError.sdkNotInitialized)
            return
        }
        let beaconContextIds = controller.getNearbyBeacons()
        var contextsArray = JSArray()
        if let beaconContextIds = beaconContextIds {
            for contextId in beaconContextIds {
                if let context = controller.getContext(contextId) {
                    contextsArray.append(PixliveHelper.contextToJSObject(context))
                }
            }
        }
        completion(GetNearbyBeaconsResult(contexts: contextsArray), nil)
    }

    @objc public func startNearbyGPSDetection(completion: @escaping (_ error: Error?) -> Void) {
        VDARSDKController.sharedInstance()?.startNearbyGPSDetection()
        completion(nil)
    }

    @objc public func stopNearbyGPSDetection(completion: @escaping (_ error: Error?) -> Void) {
        VDARSDKController.sharedInstance()?.stopNearbyGPSDetection()
        completion(nil)
    }

    @objc public func startGPSNotifications(completion: @escaping (_ error: Error?) -> Void) {
        VDARSDKController.sharedInstance()?.startGPSNotifications()
        completion(nil)
    }

    @objc public func stopGPSNotifications(completion: @escaping (_ error: Error?) -> Void) {
        VDARSDKController.sharedInstance()?.stopGPSNotifications()
        completion(nil)
    }

    @objc public func setNotificationsSupport(_ options: SetNotificationsSupportOptions, completion: @escaping (_ error: Error?) -> Void) {
        VDARSDKController.sharedInstance()?.isNotificationsEnabled = options.enabled
        completion(nil)
    }

    @objc public func setInterfaceLanguage(_ options: SetInterfaceLanguageOptions, completion: @escaping (_ error: Error?) -> Void) {
        VDARSDKController.sharedInstance()?.forceLanguage(options.language)
        completion(nil)
    }

    @objc public func createARView(_ options: CreateARViewOptions, completion: @escaping (_ error: Error?) -> Void) {
        if arViewController != nil {
            completion(CustomError.arViewAlreadyExists)
            return
        }
        guard let plugin = plugin else { return }
        DispatchQueue.main.async {
            guard let webView = plugin.webView, let superview = webView.superview else { return }
            let arVC = CapacitorARViewController()
            arVC.pixlive = self
            self.arViewController = arVC
            _ = arVC.view
            arVC.viewDidLoad()
            let frame = CGRect(x: options.x, y: options.y, width: options.width, height: options.height)
            arVC.view.frame = frame
            webView.isOpaque = false
            webView.backgroundColor = .clear
            webView.scrollView.backgroundColor = .clear

            let forwarder = TouchForwarderView(frame: webView.frame)
            forwarder.autoresizingMask = webView.autoresizingMask
            superview.insertSubview(forwarder, aboveSubview: webView)
            webView.removeFromSuperview()
            forwarder.addSubview(webView)
            webView.frame = forwarder.bounds
            webView.autoresizingMask = [.flexibleWidth, .flexibleHeight]

            forwarder.arView = arVC.view
            self.touchForwarderView = forwarder

            superview.insertSubview(arVC.view, belowSubview: forwarder)
            arVC.viewWillAppear(false)
            arVC.viewDidAppear(false)
            completion(nil)
        }
    }

    @objc public func destroyARView(completion: @escaping (_ error: Error?) -> Void) {
        guard let arVC = arViewController else {
            completion(CustomError.arViewNotFound)
            return
        }
        DispatchQueue.main.async {
            guard let plugin = self.plugin, let webView = plugin.webView else { return }
            webView.isOpaque = true
            webView.backgroundColor = .white
            webView.scrollView.backgroundColor = .white
            arVC.view.removeFromSuperview()
            arVC.stopAndUnload()
            self.arViewController = nil

            if let forwarder = self.touchForwarderView, let superview = forwarder.superview {
                webView.removeFromSuperview()
                let savedFrame = forwarder.frame
                let savedMask = forwarder.autoresizingMask
                superview.insertSubview(webView, aboveSubview: forwarder)
                forwarder.removeFromSuperview()
                webView.frame = savedFrame
                webView.autoresizingMask = savedMask
                self.touchForwarderView = nil
            }

            completion(nil)
        }
    }

    @objc public func resizeARView(_ options: ResizeARViewOptions, completion: @escaping (_ error: Error?) -> Void) {
        guard let arVC = arViewController else {
            completion(CustomError.arViewNotFound)
            return
        }
        DispatchQueue.main.async {
            let frame = CGRect(x: options.x, y: options.y, width: options.width, height: options.height)
            arVC.view.frame = frame
            completion(nil)
        }
    }

    @objc public func setARViewTouchEnabled(_ options: SetARViewTouchEnabledOptions, completion: @escaping (_ error: Error?) -> Void) {
        DispatchQueue.main.async {
            self.touchForwarderView?.touchEnabled = options.enabled
        }
        completion(nil)
    }

    @objc public func setARViewTouchHole(_ options: SetARViewTouchHoleOptions, completion: @escaping (_ error: Error?) -> Void) {
        let hole = CGRect(
            x: options.left,
            y: options.top,
            width: options.right - options.left,
            height: options.bottom - options.top
        )
        DispatchQueue.main.async {
            self.touchForwarderView?.touchHole = hole
        }
        completion(nil)
    }

    // MARK: - CapacitorARViewController

    func annotationViewDidBecomeEmpty() {
        plugin?.notifyListenersFromImplementation("hideAnnotations", data: JSObject())
    }

    func annotationViewDidPresentAnnotations() {
        plugin?.notifyListenersFromImplementation("presentAnnotations", data: JSObject())
    }

    // MARK: - VDARSDKControllerDelegate

    // swiftlint:disable:next implicitly_unwrapped_optional
    @objc public func didEnter(_ context: VDARContext!) {
        guard let context = context else { return }
        self.currentContext = context
        var data = JSObject()
        data["contextId"] = context.remoteID ?? ""
        plugin?.notifyListenersFromImplementation("enterContext", data: data)
    }

    // swiftlint:disable:next implicitly_unwrapped_optional
    @objc public func didExit(_ context: VDARContext!) {
        guard let context = context else { return }
        self.currentContext = nil
        var data = JSObject()
        data["contextId"] = context.remoteID ?? ""
        plugin?.notifyListenersFromImplementation("exitContext", data: data)
    }

    // swiftlint:disable:next implicitly_unwrapped_optional
    @objc public func codesDetected(_ codes: [VDARCode]!) {
        guard let codes = codes else { return }
        for code in codes {
            if code.isSpecialCode {
                continue
            }
            var data = JSObject()
            data["code"] = code.codeData ?? ""
            data["type"] = PixliveHelper.codeTypeToString(code.codeType)
            plugin?.notifyListenersFromImplementation("codeRecognize", data: data)
        }
    }

    // swiftlint:disable:next implicitly_unwrapped_optional
    @objc public func errorOccurred(onModelManager err: Error!) {
        // Required delegate method - no action needed
    }

    // swiftlint:disable:next implicitly_unwrapped_optional
    @objc public func didDispatchEvent(inApp eventName: String!, withParams eventParams: String!) {
        guard let eventName = eventName else { return }
        var data = JSObject()
        data["name"] = eventName
        data["params"] = eventParams ?? ""
        plugin?.notifyListenersFromImplementation("eventFromContent", data: data)
    }

    // swiftlint:disable:next implicitly_unwrapped_optional
    @objc public func contextDidRequireSynchronization(_ priors: [VDARPrior]!) {
        guard let priors = priors else { return }
        var tags: [String] = []
        for prior in priors {
            if let tagPrior = prior as? VDARTagPrior {
                tags.append(tagPrior.tagName)
            }
        }
        var data = JSObject()
        data["tags"] = tags
        plugin?.notifyListenersFromImplementation("requireSync", data: data)
    }

    // MARK: - RemoteControllerDelegate

    // swiftlint:disable:next implicitly_unwrapped_optional
    @objc public func remoteController(_ controller: VDARRemoteController!, didProgress prc: Float, isReady: Bool, folder: String!) {
        var data = JSObject()
        data["progress"] = prc
        plugin?.notifyListenersFromImplementation("syncProgress", data: data)
    }
}

File: packages/pixlive/android/src/main/java/io/capawesome/capacitorjs/plugins/pixlive/Pixlive.java
package io.capawesome.capacitorjs.plugins.pixlive;

import android.app.Activity;
import android.graphics.Color;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebView;
import android.widget.FrameLayout;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.vidinoti.android.vdarsdk.VDARAnnotationView;
import com.vidinoti.android.vdarsdk.VDARCode;
import com.vidinoti.android.vdarsdk.VDARContentEventReceiver;
import com.vidinoti.android.vdarsdk.VDARContext;
import com.vidinoti.android.vdarsdk.VDARPrior;
import com.vidinoti.android.vdarsdk.VDARRemoteController;
import com.vidinoti.android.vdarsdk.VDARRemoteControllerListener;
import com.vidinoti.android.vdarsdk.VDARSDKController;
import com.vidinoti.android.vdarsdk.VDARSDKControllerEventReceiver;
import com.vidinoti.android.vdarsdk.VDARTagPrior;
import com.vidinoti.android.vdarsdk.camera.DeviceCameraImageSender;
import com.vidinoti.android.vdarsdk.geopoint.GeoPointManager;
import com.vidinoti.android.vdarsdk.geopoint.VDARGPSPoint;
import io.capawesome.capacitorjs.plugins.pixlive.classes.CustomExceptions;
import io.capawesome.capacitorjs.plugins.pixlive.classes.options.*;
import io.capawesome.capacitorjs.plugins.pixlive.classes.results.*;
import io.capawesome.capacitorjs.plugins.pixlive.interfaces.EmptyCallback;
import io.capawesome.capacitorjs.plugins.pixlive.interfaces.NonEmptyResultCallback;
import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Observable;
import java.util.Observer;

public class Pixlive implements VDARSDKControllerEventReceiver, VDARContentEventReceiver, VDARRemoteControllerListener {

    @NonNull
    private final PixlivePlugin plugin;

    @Nullable
    private VDARAnnotationView annotationView;

    @Nullable
    private TouchInterceptorView touchInterceptorView;

    @Nullable
    private VDARContext currentContext;

    public Pixlive(@NonNull PixlivePlugin plugin) {
        this.plugin = plugin;
    }

    public void initialize() {
        Activity activity = plugin.getActivity();
        String licenseKey = plugin.getConfig().getString("licenseKey", "");
        String apiUrl = plugin.getConfig().getString("apiUrl", null);
        String sdkUrl = plugin.getConfig().getString("sdkUrl", null);

        File storageDir = new File(activity.getFilesDir(), "pixliveSDK");
        if (!storageDir.exists()) {
            storageDir.mkdirs();
        }

        VDARSDKController.startSDK(activity, storageDir.getAbsolutePath(), licenseKey);
        VDARSDKController controller = VDARSDKController.getInstance();
        if (apiUrl != null) {
            VDARRemoteController.getInstance().setCustomRemoteApiServerEndpoint(apiUrl);
        }
        if (sdkUrl != null) {
            VDARRemoteController.getInstance().setCustomRemoteSdkApiServerEndpoint(sdkUrl);
        }
        controller.setEnableCodesRecognition(true);
        controller.setActivity(activity);
        try {
            controller.setImageSender(new DeviceCameraImageSender());
        } catch (Exception e) {
            // Camera not available
        }
        controller.registerEventReceiver(this);
        controller.registerContentEventReceiver(this);
        VDARRemoteController.getInstance().addProgressListener(this);
    }

    public void handleOnPause() {
        if (annotationView != null && annotationView.getParent() != null && annotationView.getVisibility() == View.VISIBLE) {
            annotationView.onPause();
        }
        VDARSDKController controller = VDARSDKController.getInstance();
        if (controller != null) {
            controller.onActivityPaused(plugin.getActivity());
        }
    }

    public void handleOnResume() {
        if (annotationView != null && annotationView.getParent() != null && annotationView.getVisibility() == View.VISIBLE) {
            annotationView.onResume();
        }
        VDARSDKController controller = VDARSDKController.getInstance();
        if (controller != null) {
            controller.onActivityResumed(plugin.getActivity());
        }
    }

    public void handleOnDestroy() {
        VDARSDKController controller = VDARSDKController.getInstance();
        if (controller != null) {
            controller.unregisterEventReceiver(this);
            controller.unregisterContentEventReceiver(this);
            VDARRemoteController.getInstance().removeProgressListener(this);
        }
    }

    public void synchronize(@NonNull SynchronizeOptions options, @NonNull EmptyCallback callback) {
        try {
            List<VDARPrior> priors = PixliveHelper.buildTagPriors(options.getTags());
            ArrayList<VDARPrior> priorsList = new ArrayList<>(priors);
            VDARRemoteController.getInstance()
                .syncRemoteContextsAsynchronouslyWithPriors(
                    priorsList,
                    new Observer() {
                        @Override
                        public void update(Observable o, Object arg) {
                            VDARRemoteController.ObserverUpdateInfo info = (VDARRemoteController.ObserverUpdateInfo) arg;
                            if (!info.isCompleted()) {
                                return;
                            }
                            if (info.getError() != null) {
                                callback.error(new Exception(info.getError()));
                            } else {
                                callback.success();
                            }
                        }
                    }
                );
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void synchronizeWithToursAndContexts(@NonNull SynchronizeWithToursAndContextsOptions options, @NonNull EmptyCallback callback) {
        try {
            List<VDARPrior> priors = PixliveHelper.buildFullPriors(options.getTags(), options.getTourIds(), options.getContextIds());
            ArrayList<VDARPrior> priorsList = new ArrayList<>(priors);
            VDARRemoteController.getInstance()
                .syncRemoteContextsAsynchronouslyWithPriors(
                    priorsList,
                    new Observer() {
                        @Override
                        public void update(Observable o, Object arg) {
                            VDARRemoteController.ObserverUpdateInfo info = (VDARRemoteController.ObserverUpdateInfo) arg;
                            if (!info.isCompleted()) {
                                return;
                            }
                            if (info.getError() != null) {
                                callback.error(new Exception(info.getError()));
                            } else {
                                callback.success();
                            }
                        }
                    }
                );
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void updateTagMapping(@NonNull UpdateTagMappingOptions options, @NonNull EmptyCallback callback) {
        try {
            List<String> tags = new ArrayList<>();
            for (int i = 0; i < options.getTags().length(); i++) {
                tags.add(options.getTags().getString(i));
            }
            VDARRemoteController.getInstance()
                .syncTagContexts(
                    tags,
                    new VDARRemoteController.Callback<Void>() {
                        @Override
                        public void onSuccess(Void result) {
                            callback.success();
                        }

                        @Override
                        public void onError(String message, Throwable throwable) {
                            callback.error(new Exception(message, throwable));
                        }
                    }
                );
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void enableContextsWithTags(@NonNull EnableContextsWithTagsOptions options, @NonNull EmptyCallback callback) {
        try {
            List<String> tags = new ArrayList<>();
            for (int i = 0; i < options.getTags().length(); i++) {
                tags.add(options.getTags().getString(i));
            }
            VDARSDKController.getInstance().disableContexts();
            VDARSDKController.getInstance().enableContextsWithTags(tags);
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void getContexts(@NonNull NonEmptyResultCallback<GetContextsResult> callback) {
        try {
            ArrayList<String> contextIds = VDARSDKController.getInstance().getAllContextIDs();
            JSArray contextsArray = new JSArray();
            if (contextIds != null) {
                for (String contextId : contextIds) {
                    VDARContext context = VDARSDKController.getInstance().getContext(contextId);
                    if (context != null) {
                        contextsArray.put(PixliveHelper.contextToJSObject(context));
                    }
                }
            }
            callback.success(new GetContextsResult(contextsArray));
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void getContext(@NonNull GetContextOptions options, @NonNull NonEmptyResultCallback<GetContextResult> callback) {
        try {
            VDARContext context = VDARSDKController.getInstance().getContext(options.getContextId());
            if (context == null) {
                callback.error(CustomExceptions.CONTEXT_NOT_FOUND);
                return;
            }
            callback.success(new GetContextResult(PixliveHelper.contextToJSObject(context)));
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void activateContext(@NonNull ActivateContextOptions options, @NonNull EmptyCallback callback) {
        try {
            VDARContext context = VDARSDKController.getInstance().getContext(options.getContextId());
            if (context != null) {
                context.activate();
            }
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void stopContext(@NonNull EmptyCallback callback) {
        try {
            if (currentContext != null) {
                currentContext.stop();
            }
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void getNearbyGPSPoints(
        @NonNull GetNearbyGPSPointsOptions options,
        @NonNull NonEmptyResultCallback<GetNearbyGPSPointsResult> callback
    ) {
        try {
            List<VDARGPSPoint> points = GeoPointManager.getNearbyGPSPoints((float) options.getLatitude(), (float) options.getLongitude());
            JSArray pointsArray = new JSArray();
            if (points != null) {
                for (VDARGPSPoint point : points) {
                    pointsArray.put(PixliveHelper.gpsPointToJSObject(point));
                }
            }
            callback.success(new GetNearbyGPSPointsResult(pointsArray));
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void getGPSPointsInBoundingBox(
        @NonNull GetGPSPointsInBoundingBoxOptions options,
        @NonNull NonEmptyResultCallback<GetGPSPointsInBoundingBoxResult> callback
    ) {
        try {
            List<VDARGPSPoint> points = GeoPointManager.getGPSPointsInBoundingBox(
                (float) options.getMinLatitude(),
                (float) options.getMinLongitude(),
                (float) options.getMaxLatitude(),
                (float) options.getMaxLongitude()
            );
            JSArray pointsArray = new JSArray();
            if (points != null) {
                for (VDARGPSPoint point : points) {
                    pointsArray.put(PixliveHelper.gpsPointToJSObject(point));
                }
            }
            callback.success(new GetGPSPointsInBoundingBoxResult(pointsArray));
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void getNearbyBeacons(@NonNull NonEmptyResultCallback<GetNearbyBeaconsResult> callback) {
        try {
            List<String> beaconContextIds = VDARSDKController.getInstance().getNearbyBeacons();
            JSArray contextsArray = new JSArray();
            if (beaconContextIds != null) {
                for (String contextId : beaconContextIds) {
                    VDARContext context = VDARSDKController.getInstance().getContext(contextId);
                    if (context != null) {
                        contextsArray.put(PixliveHelper.contextToJSObject(context));
                    }
                }
            }
            callback.success(new GetNearbyBeaconsResult(contextsArray));
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void startNearbyGPSDetection(@NonNull EmptyCallback callback) {
        try {
            VDARSDKController.getInstance().startNearbyGPSDetection();
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void stopNearbyGPSDetection(@NonNull EmptyCallback callback) {
        try {
            VDARSDKController.getInstance().stopNearbyGPSDetection();
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void startGPSNotifications(@NonNull EmptyCallback callback) {
        try {
            VDARSDKController.getInstance().startGPSNotifications();
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void stopGPSNotifications(@NonNull EmptyCallback callback) {
        try {
            VDARSDKController.getInstance().stopGPSNotifications();
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void setNotificationsSupport(@NonNull SetNotificationsSupportOptions options, @NonNull EmptyCallback callback) {
        try {
            VDARSDKController.getInstance().setNotificationsSupport(options.isEnabled());
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void setInterfaceLanguage(@NonNull SetInterfaceLanguageOptions options, @NonNull EmptyCallback callback) {
        try {
            VDARSDKController.getInstance().forceLanguage(options.getLanguage());
            callback.success();
        } catch (Exception exception) {
            callback.error(exception);
        }
    }

    public void createARView(@NonNull CreateARViewOptions options, @NonNull EmptyCallback callback) {
        if (annotationView != null) {
            callback.error(CustomExceptions.AR_VIEW_ALREADY_EXISTS);
            return;
        }
        Activity activity = plugin.getActivity();
        activity.runOnUiThread(() -> {
            try {
                WebView webView = plugin.getBridge().getWebView();
                annotationView = new VDARAnnotationView(activity);
                float density = activity.getResources().getDisplayMetrics().density;
                int x = (int) (options.getX() * density);
                int y = (int) (options.getY() * density);
                int width = (int) (options.getWidth() * density);
                int height = (int) (options.getHeight() * density);
                FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(width, height);
                params.leftMargin = x;
                params.topMargin = y;
                annotationView.setLayoutParams(params);
                webView.setBackgroundColor(Color.TRANSPARENT);

                ViewGroup parent = (ViewGroup) webView.getParent();
                int index = parent.indexOfChild(webView);
                ViewGroup.LayoutParams webViewParams = webView.getLayoutParams();
                parent.removeView(webView);

                TouchInterceptorView interceptor = new TouchInterceptorView(activity);
                interceptor.setLayoutParams(webViewParams);
                parent.addView(interceptor, index);

                webView.setLayoutParams(
                    new FrameLayout.LayoutParams(FrameLayout.LayoutParams.MATCH_PARENT, FrameLayout.LayoutParams.MATCH_PARENT)
                );
                interceptor.addView(webView);
                interceptor.addView(annotationView, 0);
                interceptor.setAnnotationView(annotationView);
                touchInterceptorView = interceptor;

                VDARSDKController.getInstance().setActivity(activity);
                annotationView.onResume();
                callback.success();
            } catch (Exception exception) {
                callback.error(exception);
            }
        });
    }

    public void destroyARView(@NonNull EmptyCallback callback) {
        if (annotationView == null) {
            callback.error(CustomExceptions.AR_VIEW_NOT_FOUND);
            return;
        }
        Activity activity = plugin.getActivity();
        activity.runOnUiThread(() -> {
            try {
                WebView webView = plugin.getBridge().getWebView();
                webView.setBackgroundColor(Color.WHITE);
                if (currentContext != null) {
                    currentContext.stop();
                    currentContext = null;
                }
                annotationView.onPause();

                if (touchInterceptorView != null) {
                    ViewGroup parent = (ViewGroup) touchInterceptorView.getParent();
                    int index = parent.indexOfChild(touchInterceptorView);
                    ViewGroup.LayoutParams interceptorParams = touchInterceptorView.getLayoutParams();

                    touchInterceptorView.removeView(annotationView);
                    touchInterceptorView.removeView(webView);
                    parent.removeView(touchInterceptorView);

                    webView.setLayoutParams(interceptorParams);
                    parent.addView(webView, index);
                    touchInterceptorView = null;
                }

                annotationView = null;
                callback.success();
            } catch (Exception exception) {
                callback.error(exception);
            }
        });
    }

    public void resizeARView(@NonNull ResizeARViewOptions options, @NonNull EmptyCallback callback) {
        if (annotationView == null) {
            callback.error(CustomExceptions.AR_VIEW_NOT_FOUND);
            return;
        }
        Activity activity = plugin.getActivity();
        activity.runOnUiThread(() -> {
            try {
                float density = activity.getResources().getDisplayMetrics().density;
                int x = (int) (options.getX() * density);
                int y = (int) (options.getY() * density);
                int width = (int) (options.getWidth() * density);
                int height = (int) (options.getHeight() * density);
                FrameLayout.LayoutParams params = new FrameLayout.LayoutParams(width, height);
                params.leftMargin = x;
                params.topMargin = y;
                annotationView.setLayoutParams(params);
                callback.success();
            } catch (Exception exception) {
                callback.error(exception);
            }
        });
    }

    public void setARViewTouchEnabled(@NonNull SetARViewTouchEnabledOptions options, @NonNull EmptyCallback callback) {
        if (touchInterceptorView != null) {
            plugin
                .getActivity()
                .runOnUiThread(() -> {
                    if (touchInterceptorView != null) {
                        touchInterceptorView.setTouchEnabled(options.isEnabled());
                    }
                });
        }
        callback.success();
    }

    public void setARViewTouchHole(@NonNull SetARViewTouchHoleOptions options, @NonNull EmptyCallback callback) {
        float density = plugin.getActivity().getResources().getDisplayMetrics().density;
        float top = (float) (options.getTop() * density);
        float bottom = (float) (options.getBottom() * density);
        float left = (float) (options.getLeft() * density);
        float right = (float) (options.getRight() * density);
        if (touchInterceptorView != null) {
            plugin
                .getActivity()
                .runOnUiThread(() -> {
                    if (touchInterceptorView != null) {
                        touchInterceptorView.setTouchHole(top, bottom, left, right);
                    }
                });
        }
        callback.success();
    }

    // VDARSDKControllerEventReceiver

    @Override
    public void onCodesRecognized(@NonNull ArrayList<VDARCode> codes) {
        for (VDARCode code : codes) {
            if (code.isSpecialCode()) {
                continue;
            }
            JSObject data = new JSObject();
            data.put("code", code.getCodeData());
            data.put("type", PixliveHelper.codeTypeToString(code.getCodeType()));
            plugin.notifyListenersFromImplementation("codeRecognize", data);
        }
    }

    @Override
    public void onFatalError(@NonNull String error) {
        // Not needed
    }

    @Override
    public void onPresentAnnotations() {
        plugin.notifyListenersFromImplementation("presentAnnotations", new JSObject());
    }

    @Override
    public void onAnnotationsHidden() {
        plugin.notifyListenersFromImplementation("hideAnnotations", new JSObject());
    }

    @Override
    public void onTrackingStarted(int width, int height) {
        // Not needed
    }

    @Override
    public void onEnterContext(@NonNull VDARContext context) {
        this.currentContext = context;
        JSObject data = new JSObject();
        data.put("contextId", context.getRemoteID());
        plugin.notifyListenersFromImplementation("enterContext", data);
    }

    @Override
    public void onExitContext(@NonNull VDARContext context) {
        this.currentContext = null;
        JSObject data = new JSObject();
        data.put("contextId", context.getRemoteID());
        plugin.notifyListenersFromImplementation("exitContext", data);
    }

    @Override
    public void onRequireSynchronization(@NonNull ArrayList<VDARPrior> priors) {
        JSArray tagsArray = new JSArray();
        for (VDARPrior prior : priors) {
            if (prior instanceof VDARTagPrior) {
                tagsArray.put(((VDARTagPrior) prior).getTag());
            }
        }
        JSObject data = new JSObject();
        data.put("tags", tagsArray);
        plugin.notifyListenersFromImplementation("requireSync", data);
    }

    // VDARContentEventReceiver

    @Override
    public void onReceiveContentEvent(@NonNull String eventName, @NonNull String eventParams) {
        JSObject data = new JSObject();
        data.put("name", eventName);
        data.put("params", eventParams);
        plugin.notifyListenersFromImplementation("eventFromContent", data);
    }

    // VDARRemoteControllerListener

    @Override
    public void onSyncProgress(@NonNull VDARRemoteController controller, float progress, boolean isDone, @Nullable String error) {
        JSObject data = new JSObject();
        data.put("progress", progress / 100.0f);
        plugin.notifyListenersFromImplementation("syncProgress", data);
    }
}
